
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Welcome to detectpolya’s documentation! &#8212; detectpolya 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="welcome-to-detectpolya-s-documentation">
<h1>Welcome to detectpolya’s documentation!<a class="headerlink" href="#welcome-to-detectpolya-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
<dl class="function">
<dt id="detectpolya.detection_algorithms.detectPolyA">
<code class="descclassname">detectpolya.detection_algorithms.</code><code class="descname">detectPolyA</code><span class="sig-paren">(</span><em>seq</em>, <em>qual=None</em>, <em>method='seed'</em>, <em>min_len=5</em>, <em>max_prop_non_a=0.2</em>, <em>seed_len=4</em><span class="sig-paren">)</span><a class="headerlink" href="#detectpolya.detection_algorithms.detectPolyA" title="Permalink to this definition">¶</a></dt>
<dd><p>Detects poly-adenylation in a read sequence.</p>
<p>If quality string is specified (<cite>qual</cite>) computes the probability that a
nucleotide is an adenosines and uses the expected number of adenosines in
a subsequence instead of the exact count of called adenosines. Only
returns longest match if it is more than a certain length.</p>
<ul class="simple">
<li><dl class="first docutils">
<dt><cite>seed</cite>: Heuristic similar to the BLAST algorithm. First matches a seed</dt>
<dd>subsequence of consecutive adenosines, then tries to extend it until until
a the subsequence reaches the max proportion of non-adenosines nucleotides
(<cite>max_prop_non_a</cite>);</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><cite>window</cite>: Exact method that recursively computes the proportion of</dt>
<dd>adenosines for all possible subsequences. Returns longest match that is
not below the max proportion of non-adenosines nucleotides if any.</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><cite>mixed</cite>: Tries to match a seed of consecutive adenosines. If seed is found</dt>
<dd>uses window method to identify position. If not, is ignored. This allows
to ignore unlikely matches.</dd>
</dl>
</li>
</ul>
<p>Looks for match in sequence and complement (A or T).</p>
<p class="rubric">Notes</p>
<p>The quality string is not always in the same order as the sequence.
This is true in the BAM file where the sequence can be reversed
complemented to be aligned to reference. Quality string needs to be
reversed in this case.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>seq</strong> (<em>str</em>) – Read sequence string.</li>
<li><strong>qual</strong> (<em>str</em>) – Quality string.</li>
<li><strong>method</strong> (<em>str</em>) – Detection algorithm can be <cite>seed</cite> or <cite>window</cite>.</li>
<li><strong>min_len</strong> (<em>int</em>) – Minimum length of a poly-adelynated tail.</li>
<li><strong>max_prop_non_a</strong> (<em>float</em>) – Maximum proportion of non-adenosines a
poly-adelynated tail may contain.</li>
<li><strong>seed_len</strong> (<em>int</em>) – Length of seed for seed algorithm.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><dl class="docutils">
<dt>collection.namedtuple</dt>
<dd><p class="first last">If a match is found, return a named tupple: start, end, score, strand.
The score corresponds to the number of (expected) matched adenosines.</p>
</dd>
</dl>
</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">polya</span> <span class="o">=</span> <span class="n">detectpolya</span><span class="o">.</span><span class="n">detectPolyA</span><span class="p">(</span><span class="s2">&quot;ACTGGTAAAAAA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">polya</span><span class="p">)</span>
<span class="go">Match(start=5, end=12, score=6.0, strand=&#39;+&#39;)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">polya</span> <span class="o">=</span> <span class="n">detectpolya</span><span class="o">.</span><span class="n">detectPolyA</span><span class="p">(</span><span class="s2">&quot;ACTGGTGTACAT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">polya</span><span class="p">)</span>
<span class="go">None</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="detectpolya.detection_algorithms.detectSubSequence">
<code class="descclassname">detectpolya.detection_algorithms.</code><code class="descname">detectSubSequence</code><span class="sig-paren">(</span><em>seq</em>, <em>subseq</em>, <em>min_len=7</em>, <em>max_l_dist=1</em><span class="sig-paren">)</span><a class="headerlink" href="#detectpolya.detection_algorithms.detectSubSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Uses the <cite>taleinat/fuzzysearch</cite> package to look for a subsequence within a
sequence while allowing for mismatches. Looks for subsequence itself and
its reverse complement. Only returns longest match if it is  more than a
certain length.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>seq</strong> (<em>str</em>) – Read sequence string.</li>
<li><strong>subseq</strong> (<em>str</em>) – Subsequence to be detected.</li>
<li><strong>min_len</strong> (<em>int</em>) – Minimum length of subsequence match.</li>
<li><strong>max_l_dist</strong> (<em>float</em>) – Maximum Levenshtein distance.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><dl class="docutils">
<dt>collection.namedtuple</dt>
<dd><p class="first last">If a match is found, return a named tupple: start, end, score, strand.
The score corresponds to the number of matched nucleotides.</p>
</dd>
</dl>
</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">match</span> <span class="o">=</span> <span class="n">detectpolya</span><span class="o">.</span><span class="n">detectSubSequence</span><span class="p">(</span><span class="s2">&quot;AAATATAAATACCC&quot;</span><span class="p">,</span> <span class="s2">&quot;TATATATA&quot;</span><span class="p">);</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">match</span><span class="p">)</span>
<span class="go">Match(start=3, end=10, score=6, strand=&#39;+&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="detectpolya.file_manipulation.analyseFile">
<code class="descclassname">detectpolya.file_manipulation.</code><code class="descname">analyseFile</code><span class="sig-paren">(</span><em>filename</em>, <em>filetype='fa'</em>, <em>paired_ends=False</em>, <em>gtf_filename=None</em>, <em>polya_method='seed'</em>, <em>polya_min_len=5</em>, <em>polya_max_prop_non_a=0.2</em>, <em>polya_seed_len=3</em>, <em>primer_seq=None</em>, <em>primer_min_len=10</em>, <em>primer_max_dist=1</em>, <em>nlines=None</em>, <em>gtf_nlines=None</em>, <em>verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#detectpolya.file_manipulation.analyseFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Detects poly-adenylation in a file containing reads. Supports FASTA, FASTQ, SAM and BAM</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>filename</strong> (<em>str</em>) – Path of file to be analyzed. For paired ended fasta and fastq, must be a list of filenames.</li>
<li><strong>qual</strong> (<em>str</em>) – Type of file. Can be “fa” for fasta, “fq” for fastq, “sam” or “bam”.</li>
<li><strong>paired_ends</strong> (<em>bool</em>) – Was the experiment paired-ended?</li>
<li><strong>gtf_filename</strong> (<em>str</em>) – GTF filename.</li>
<li><strong>polya_method</strong> (<em>str</em>) – Detection algorithm can be <cite>seed</cite> or <cite>window</cite>.</li>
<li><strong>polya_min_len</strong> (<em>int</em>) – Minimum length of a poly-adelynated tail.</li>
<li><strong>polya_max_prop_non_a</strong> (<em>float</em>) – Maximum proportion of non-adenosines a poly-adelynated tail may contain.</li>
<li><strong>polya_seed_len</strong> (<em>int</em>) – Length of seed for seed algorithm.</li>
<li><strong>primer_seq</strong> (<em>str</em>) – Primer sequence to be detected.</li>
<li><strong>primer_min_len</strong> (<em>int</em>) – Minimum length of primer.</li>
<li><strong>primer_max_l_dist</strong> (<em>float</em>) – Maximum Levenshtein distance; mismatched nucleotides between reference and primer.</li>
<li><strong>nlines</strong> (<em>int</em>) – Maximum number of lines to be read.</li>
<li><strong>gtf_nlines</strong> (<em>int</em>) – Maximum number of lines to be read in GTF file.</li>
<li><strong>verbose</strong> (<em>bool</em>) – Print progress</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><dl class="docutils">
<dt>dict</dt>
<dd><p class="first last">Keys are Gene IDs as found in the GTF file, if specified.
If no GTF was specified, there is only one key: “_unmapped”.
Values is a list of dictionnaries.
Every dictionnary correspond to a read with a poly-A or primer detected.
The keys are described in README.</p>
</dd>
</dl>
</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">results</span> <span class="o">=</span> <span class="n">detectpolya</span><span class="o">.</span><span class="n">analyseFile</span><span class="p">(</span><span class="s2">&quot;./test/paired_ends.sam&quot;</span><span class="p">,</span> <span class="s2">&quot;sam&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
<span class="go">defaultdict(list,</span>
<span class="go">        {&#39;_unmapped&#39;: [{&#39;gene_id&#39;: &#39;_unmapped&#39;,</span>
<span class="go">           &#39;polya_end_in_genome&#39;: &#39;15063255&#39;,</span>
<span class="go">           &#39;polya_end_in_read&#39;: &#39;72&#39;,</span>
<span class="go">           &#39;polya_length&#39;: &#39;21&#39;,</span>
<span class="go">           &#39;polya_score&#39;: &#39;16.9969746427&#39;,</span>
<span class="go">           &#39;polya_start_in_genome&#39;: &#39;15063184&#39;,</span>
<span class="go">           &#39;polya_start_in_read&#39;: &#39;52&#39;,</span>
<span class="go">           &#39;primer_end_in_genome&#39;: None,</span>
<span class="go">           &#39;primer_end_in_read&#39;: None,</span>
<span class="go">           &#39;primer_length&#39;: None,</span>
<span class="go">           &#39;primer_score&#39;: None,</span>
<span class="go">           &#39;primer_start_in_genome&#39;: None,</span>
<span class="go">           &#39;primer_start_in_read&#39;: None,</span>
<span class="go">           &#39;read_chrom&#39;: &#39;I&#39;,</span>
<span class="go">           &#39;read_cigar&#39;: &#39;51M25S&#39;,</span>
<span class="go">           &#39;read_clipped_seq&#39;: &#39;===================================================AAAAAAAAAAAAAAAAGTACTCTGC&#39;,</span>
<span class="go">           &#39;read_count&#39;: 18,</span>
<span class="go">           &#39;read_end&#39;: 15063184,</span>
<span class="go">           &#39;read_length&#39;: 51,</span>
<span class="go">           &#39;read_mate&#39;: None,</span>
<span class="go">           &#39;read_name&#39;: &#39;D00224L:232:CCB68ANXX:4:1101:14325:85615&#39;,</span>
<span class="go">           &#39;read_qual&#39;: &#39;0000&lt;&lt;7&lt;707070FBB&lt;00000000000&lt;0FB&lt;0007BBFFFFFFFB&lt;0FFFFFFFFFFFFFFFFFFFFFBB&lt;&lt;0&#39;,</span>
<span class="go">           &#39;read_reversed_complemented&#39;: True,</span>
<span class="go">           &#39;read_seq&#39;: &#39;AGTTTTTCGTTTCCGGGGGTAGTATGGTTGAAAAGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGTACTCTGC&#39;,</span>
<span class="go">           &#39;read_start&#39;: 15063133,</span>
<span class="go">           &#39;transcript_end&#39;: None,</span>
<span class="go">           &#39;transcript_length&#39;: None,</span>
<span class="go">           &#39;transcript_start&#39;: None}]})</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="detectpolya.file_manipulation.printResults">
<code class="descclassname">detectpolya.file_manipulation.</code><code class="descname">printResults</code><span class="sig-paren">(</span><em>results</em>, <em>outf=None</em>, <em>header=True</em><span class="sig-paren">)</span><a class="headerlink" href="#detectpolya.file_manipulation.printResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Write results from analyseFile to a file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>results</strong> (<em>dict</em>) – Output of analyseFile</li>
<li><strong>outf</strong> (<em>file</em>) – Filed open for writing. If not specified, results are printed.</li>
<li><strong>header</strong> (<em>bool</em>) – Should header be printed?</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">None</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">results</span> <span class="o">=</span> <span class="n">detectpolya</span><span class="o">.</span><span class="n">analyseFile</span><span class="p">(</span><span class="s2">&quot;./test/paired_ends.sam&quot;</span><span class="p">,</span> <span class="s2">&quot;sam&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">detectpolya</span><span class="o">.</span><span class="n">printResults</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
<span class="go">gene_id,transcript_start,transcript_end,transcript_length,read_name,read_mate,chrom,read_start,read_end,read_length,read_seq,read_clipped_seq,read_qual,read_cigar,read_reversed_complemented,read_count,polya_start_in_genome,polya_end_in_genome,polya_start_in_read,polya_end_in_read,polya_length,polya_score,primer_start_in_genome,primer_end_in_genome,primer_start_in_read,primer_end_in_read,primer_length,primer_score</span>
<span class="go">_unmapped,NA,NA,NA,D00224L:232:CCB68ANXX:4:1101:14325:85615,NA,I,15063133,15063184,51,AGTTTTTCGTTTCCGGGGGTAGTATGGTTGAAAAGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGTACTCTGC,===================================================AAAAAAAAAAAAAAAAGTACTCTGC,0000&lt;&lt;7&lt;707070FBB&lt;00000000000&lt;0FB&lt;0007BBFFFFFFFB&lt;0FFFFFFFFFFFFFFFFFFFFFBB&lt;&lt;0,51M25S,True,18,15063184,15063255,52,72,21,16.9969746427,NA,NA,NA,NA,NA,NA</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="detectpolya.misc.estimateProbabilityNucleotide">
<code class="descclassname">detectpolya.misc.</code><code class="descname">estimateProbabilityNucleotide</code><span class="sig-paren">(</span><em>seq</em>, <em>qual=None</em>, <em>nuc='A'</em><span class="sig-paren">)</span><a class="headerlink" href="#detectpolya.misc.estimateProbabilityNucleotide" title="Permalink to this definition">¶</a></dt>
<dd><p>Estimate probability of a specific nucleotide to be at a
position of a read sequence given call errors specified by
quality string.
Returns zeros and ones if quality string is absent.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>seq</strong> (<em>str</em>) – Read sequence string</li>
<li><strong>qual</strong> (<em>str</em>) – Quality string. This has to be in the same order as <cite>seq</cite>.
This is not always the case in BAM file where the sequence can be
reversed complemented to be aligned to reference. Quality string
needs to be reversed in this case.</li>
<li><strong>nuc</strong> (<em>str</em>) – nucleotide of intestest. By default, this is “A” so we can
look for poly-adelynation.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><dl class="docutils">
<dt>list</dt>
<dd><p class="first last">List of floats between zero and one giving the probability that <cite>nuc</cite>
is called at positions given by <cite>seq</cite>.</p>
</dd>
</dl>
</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">detectpolya</span><span class="o">.</span><span class="n">estimateProbabilityNucleotide</span><span class="p">(</span><span class="s2">&quot;CGTTAAATA&quot;</span><span class="p">,</span> <span class="s2">&quot;BBBCFFF!B&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="go">[0.00016706241120909083,</span>
<span class="go"> 0.00016706241120909083,</span>
<span class="go"> 0.00016706241120909083,</span>
<span class="go"> 0.0001327023901844991,</span>
<span class="go"> 0.9998004737685031,</span>
<span class="go"> 0.9998004737685031,</span>
<span class="go"> 0.9998004737685031,</span>
<span class="go"> 0.3333333333333333,</span>
<span class="go"> 0.9994988127663728]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="detectpolya.misc.removeMatches">
<code class="descclassname">detectpolya.misc.</code><code class="descname">removeMatches</code><span class="sig-paren">(</span><em>seq</em>, <em>cigar</em>, <em>remove_five_prime=False</em>, <em>remove_three_prime=False</em><span class="sig-paren">)</span><a class="headerlink" href="#detectpolya.misc.removeMatches" title="Permalink to this definition">¶</a></dt>
<dd><p>Replaces nucleotide matching reference by equal sign in read sequence.
This allows these nucleotides to be ignored by detection algorithms
while still outputting correct position in read.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>seq</strong> (<em>str</em>) – Read sequence string</li>
<li><strong>cigar</strong> (<em>str</em>) – CIGAR string</li>
<li><strong>remove_five_prime</strong> (<em>bool</em>) – Remove 5’ soft clipped nucleotides</li>
<li><strong>remove_three_prime</strong> (<em>bool</em>) – Remove 3’ soft clipped nucleotides</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><dl class="docutils">
<dt>str</dt>
<dd><p class="first last">Sequence with nucleotide matching reference replaced by equal sign.</p>
</dd>
</dl>
</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">detectpolya</span><span class="o">.</span><span class="n">removeMatches</span><span class="p">(</span><span class="s2">&quot;ACTG&quot;</span><span class="p">,</span> <span class="s2">&quot;3M1S&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">&quot;===G&quot;</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">detectpolya</span><span class="o">.</span><span class="n">removeMatches</span><span class="p">(</span><span class="s2">&quot;ACTG&quot;</span><span class="p">,</span> <span class="s2">&quot;3M1S&quot;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">&quot;A===&quot;</span>
</pre></div>
</div>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">detectpolya</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, Olivier M. F. Martin.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>